File: engine/src/main/java/com/trading/hf/core/OptionChainHandler.java
package com.trading.hf.core;

import com.lmax.disruptor.EventHandler;
import com.trading.hf.model.MarketEvent;
import com.trading.hf.model.OptionChainData;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class OptionChainHandler implements EventHandler<MarketEvent> {

    private static final Logger log = LoggerFactory.getLogger(OptionChainHandler.class);
    private static final Map<Integer, OptionChainData> latestOptionChain = new ConcurrentHashMap<>();

    @Override
    public void onEvent(MarketEvent event, long sequence, boolean endOfBatch) throws Exception {
        if (event.getType() == MarketEvent.MessageType.OPTION_CHAIN_UPDATE) {
            try {
                List<OptionChainData> chainData = event.getOptionChain();
                if (chainData != null) {
                    for (OptionChainData data : chainData) {
                        latestOptionChain.put(data.getStrike(), data);
                    }
                    log.info("Successfully processed and updated option chain data for {} strikes.", chainData.size());
                }
            } catch (Exception e) {
                log.error("Failed to process OPTION_CHAIN_UPDATE event", e);
            }
        }
    }

    public static Map<Integer, OptionChainData> getLatestOptionChain() {
        return latestOptionChain;
    }
}

File: python_engine/core/option_chain_handler.py
from typing import Dict, List
from python_engine.models.data_models import MarketEvent, OptionChainData, MessageType

class OptionChainHandler:
    def __init__(self):
        self._latest_option_chain: Dict[int, OptionChainData] = {}

    def on_event(self, event: MarketEvent):
        if event.type == MessageType.OPTION_CHAIN_UPDATE:
            if event.option_chain:
                for data in event.option_chain:
                    self._latest_option_chain[data.strike] = data

    def get_latest_option_chain(self) -> Dict[int, OptionChainData]:
        return self._latest_option_chain
